idaes/core/util/scaling.py

```diff
@@ -1041,3 +1047,165 @@ def set_derivative_factor_from_state(self, deriv, nominal_wrt=1.0):
+                missing scaling factors is encountered (default=False)
+
+        Notes
+        -----
+        This class inherits from the :class:`StreamBasedExpressionVisitor` to implement
+        a walker that returns the pyomo units and pint units corresponding to an
+        expression.
+        There are class attributes (dicts) that map the expression node type to the
+        particular method that should be called to return the units of the node based
+        on the units of its child arguments. This map is used in exitNode.
+        """
+        super().__init__()
+
+        self.warning = warning
+        self.exception = exception
+
+    def _get_scale_factor_for_sum(self, node, child_scale_factors):
+        sf = []
+        for i in child_scale_factors:
+            for j in i:
+                sf.append(j)
+        return sf
+
+    def _get_scale_factor_for_product(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 2
+
+        sf = []
+        for i in child_scale_factors[0]:
+            for j in child_scale_factors[1]:
+                sf.append(i * j)
+        return sf
+
+    def _get_scale_factor_for_division(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 2
+
+        sf = []
+        for i in child_scale_factors[0]:
+            for j in child_scale_factors[1]:
+                sf.append(i / j)
+        return sf
+
+    def _get_scale_factor_for_power(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 2
+
+        # We will create separate scaling factors for all additive terms
+        # Users can decide how to use these later
+        sf = []
+        for i in child_scale_factors[0]:
+            for j in child_scale_factors[1]:
+                sf.append(i**j)
+        return sf
+
+    def _get_scale_factor_single_child(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 1
+        return child_scale_factors[0]
+
+    def _get_scale_factor_for_unary_function(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 1
+        func_name = node.getname()
+        func = getattr(math, func_name)
+        return [func(i) for i in child_scale_factors[0]]
+
+    def _get_scale_factor_expr_if(self, node, child_scale_factors):
+        assert len(child_scale_factors) == 3
+        return child_scale_factors[1] + child_scale_factors[2]
+
+    def _get_scale_factor_no_children(self, node, child_units):
+        assert len(child_units) == 0
+        # may need more checks for dimensionless for other types
+        assert type(node) is IndexTemplate
+        return [1]
+
+    def _get_scale_factor_external_function(self, node, child_units):
+        # TODO: Work out what to do here
+        # Maybe pass child_units into the external function and get the result?
+        # For now, assume we know nothing about scaling
+        return [1]
+
+    node_type_method_map = {
```

Comment:
I think there are a few options for directly evaluating the expression with the scaling values for variables rather than the variables that should make this much much simpler.  Maybe the simplest is to swap in the scaling factor for variables values and set it back when you are done.

Reply:
Evaluating the expression with the scaling factors will fail quite frequently. If you have the expression a+b+c with a~O(1), b~O(1), and c~O(100), you have s_a = 1, s_b = 1 and s_c = 0.01. Evaluating the expression, you'd get a scaling factor of 2.01, when the expression is O(100) because c is O(100).

Reply:
Not directly evaluating the original expression, evaluating the subexpressions.

Reply:
As @eslickj  said, the result is `[1, 1, 0.01]`, not `2.01`. The aim here is to get the magnitude of all the additive terms separately, so that we can then work out what the best overall scaling would be,

Reply:
I think just to avoid confusion we should shift to expected magnitude for user-facing stuff.  Scale factors are still a little ambiguous.

Reply:
They're called scaling _factors_, not scaling _divisors_.

We want to be sure that whatever system we decide on, we're consistent between `set_scaling_factor` and `constraint_scaling_transform`

Reply:
Yeah, I know we multiply by scaling factor (I have seen them used to divide though and I agree it's counter intuitive).  It doesn't necessarily change anything for this PR, but for a final user interface would be to allow the end user to provide expected magnitude as a way to set the scaling factor.  Like we could have `set_scaling_factor` and/or `set_expected_magnitude()`.  It makes very little difference, but magnitude seems more clear and easier to thing about.

Reply:
I guess the main thing is to make sure the name is consistent with how we're using them. We can support one or both methods to set them.